pytradfri/device/light.py

```diff
@@ -24,63 +33,72 @@ class Light:
     pdf
     """
 
-    def __init__(self, device, index):
+    def __init__(self, device: Device, index: int) -> None:
         """Create object of class."""
         self.device = device
         self.index = index
 
     @property
-    def supported_features(self):
+    def supported_features(self) -> int:
         """Return supported features."""
-        return supported_features(self.raw)
+        return supported_features(cast(TYPE_RAW_SIMPLE, self.raw))
 
     @property
-    def state(self):
+    def state(self) -> bool:
         """Return device state."""
-        return self.raw.get(ATTR_DEVICE_STATE) == 1
+        return cast(TYPE_RAW_SIMPLE, self.raw).get(ATTR_DEVICE_STATE) == 1
 
     @property
-    def dimmer(self):
+    def dimmer(self) -> int | None:
         """Return dimmer if present."""
         if self.supported_features & SUPPORT_BRIGHTNESS:
-            return self.raw.get(ATTR_LIGHT_DIMMER)
+            return cast(int, cast(TYPE_RAW_SIMPLE, self.raw).get(ATTR_LIGHT_DIMMER))
+        return None
 
     @property
-    def color_temp(self):
+    def color_temp(self) -> int | None:
         """Return color temperature."""
         if self.supported_features & SUPPORT_COLOR_TEMP:
             if self.raw.get(ATTR_LIGHT_MIREDS) != 0:
-                return self.raw.get(ATTR_LIGHT_MIREDS)
+                return cast(int, cast(TYPE_RAW_SIMPLE, self.raw).get(ATTR_LIGHT_MIREDS))
+        return None
 
     @property
-    def hex_color(self):
+    def hex_color(self) -> str | None:
         """Return hex color."""
         if self.supported_features & SUPPORT_HEX_COLOR:
-            return self.raw.get(ATTR_LIGHT_COLOR_HEX)
+            return cast(str, cast(TYPE_RAW_SIMPLE, self.raw).get(ATTR_LIGHT_COLOR_HEX))
+        return None
 
     @property
-    def xy_color(self):
+    def xy_color(self) -> tuple[int, int] | None:
         """Return xy color."""
         if self.supported_features & SUPPORT_XY_COLOR:
-            return (self.raw.get(ATTR_LIGHT_COLOR_X), self.raw.get(ATTR_LIGHT_COLOR_Y))
+            return cast(
+                int, cast(TYPE_RAW_SIMPLE, self.raw).get(ATTR_LIGHT_COLOR_X)
+            ), cast(int, cast(TYPE_RAW_SIMPLE, self.raw).get(ATTR_LIGHT_COLOR_Y))
+        return None
 
     @property
-    def hsb_xy_color(self):
+    def hsb_xy_color(self) -> tuple[int, int, int, int, int]:
         """Return hsb xy color."""
         return (
-            self.raw.get(ATTR_LIGHT_COLOR_HUE),
-            self.raw.get(ATTR_LIGHT_COLOR_SATURATION),
-            self.raw.get(ATTR_LIGHT_DIMMER),
-            self.raw.get(ATTR_LIGHT_COLOR_X),
-            self.raw.get(ATTR_LIGHT_COLOR_Y),
+            cast(int, cast(TYPE_RAW_SIMPLE, self.raw).get(ATTR_LIGHT_COLOR_HUE)),
+            cast(int, cast(TYPE_RAW_SIMPLE, self.raw).get(ATTR_LIGHT_COLOR_SATURATION)),
+            cast(int, cast(TYPE_RAW_SIMPLE, self.raw).get(ATTR_LIGHT_DIMMER)),
```

Comment:
Instead of adding all these casts I think we should try to type raw better from the beginning. I think we should start in that end.

Reply:
I would love to do that, the definition (TYPE_RAW) is in resource.py

I have tried many times to limit the definition, but raw contains nearly anything.

I created TYPE_RAW_SIMPLE and TYPE_RAW__LIST as subtypes of TYPE_RAW, this needed to be able type a lot of usages e.g var = self.raw[ATTR_VAR][1], which is totally legal.

I think the right way to go is to fight through all the casts, then we have the fundament to convert raw into a proper dataclass.

Reply:
Why is that better? I don't think we gain more knowledge about raw by doing this. We should start typing at the center and move outwards.

Reply:
?? the center is typed (resource.py), I do work my way from the inside and out.

Again the problem is the all the different types of data all stored in raw. But if you have a better suggestion on to type raw I am all ears.

It is better to finalize strict typing, because then it is a lot clearer how exactly raw is used. Remember the usage is often indirected through returns in 2-3 functions.

Reply:
Raw is the response data from the gateway so that's even more central. We should type it according to what that response is not by how we use it.

Reply:
And that is what I have done in an earlier PR, the definition is in resource.py (TYPE_RAW)